File: modules/transactions/domain/transaction.ts
/**
 * `Transaction` — the canonical record of one movement of money on one
 * account, as the subject entered it or as a reviewed import committed it.
 *
 * Pure: no framework, no ORM, no clock, no randomness, no I/O. Time and
 * identity arrive as arguments (architecture test 11).
 *
 * Three properties this shape exists to guarantee:
 *
 *  1. **Exactness.** The amount is `Money` — BIGINT minor units plus a
 *     `Currency` carrying its own ISO 4217 exponent (ADR-0006). No `number`
 *     appears in any monetary position, here or in the row that stores it.
 *
 *  2. **One sign convention.** See `sign-convention.ts`. Money out of the
 *     account is negative; the source's own debit/credit wording is preserved
 *     in provenance instead of being dissolved into the sign.
 *
 *  3. **Nothing unexplainable.** Every transaction has provenance
 *     (`provenance.ts`) naming either manual input or an exact source row, and
 *     every later correction becomes a revision (`revision.ts`) rather than an
 *     overwrite.
 *
 * The three free-text fields are `HsfField` values: plaintext in memory,
 * inside a wrapper that redacts on every accidental rendering path, encrypted
 * at rest behind the `HsfFieldEncryption` port. The domain never sees a key.
 */

import type { Currency, Money } from '@karar/shared-kernel';

import type { HsfField } from './hsf-field.js';
import type { AccountRef, TransactionId } from './refs.js';

/** How the record came into existence. `CSV` covers reviewed statement import. */
export const SOURCE_KINDS = ['MANUAL', 'CSV'] as const;
export type SourceKind = (typeof SOURCE_KINDS)[number];

/**
 * Lifecycle state.
 *
 * `POSTED` is the ordinary state of a committed fact. `VOIDED` marks a record
 * a subject withdrew without deleting it — kept distinct from deletion
 * because a voided transaction still has provenance worth keeping while a
 * deleted one is gone under `CASCADE_DELETE` (MODULE.md). No state means
 * "pending review": staged rows live in the import tables the ingestion
 * workstream owns and only reach this table once committed.
 */
export const TRANSACTION_STATUSES = ['POSTED', 'VOIDED'] as const;
export type TransactionStatus = (typeof TRANSACTION_STATUSES)[number];

export class InvalidTransactionError extends Error {
  override readonly name = 'InvalidTransactionError';
}

/**
 * The amount a source stated in ITS currency, when that differs from the
 * account's — an **all-or-nothing pair**.
 *
 * Both members or neither. A currency without an amount says nothing, and an
 * amount without a currency is the classic corruption: a bare number that
 * later reads as the account currency and is wrong by a factor of the
 * exchange rate.
 *
 * **No derived exchange rate is computed or stored, ever.** Dividing the
 * booked amount by the original amount yields a number that looks like a rate
 * and is not one: it silently absorbs the institution's spread, its fees, its
 * rounding, and any partial settlement, and it changes with the exponents of
 * two currencies. Storing that quotient would create a figure this platform
 * did not observe and cannot defend, which is precisely the class of
 * fabricated financial fact the product refuses. If a source states a rate
 * explicitly, that is a source-stated fact and belongs with the import
 * evidence — still not something computed here.
 */
export interface OriginalAmount {
  readonly amount: Money;
  readonly currency: Currency;
}

export const OriginalAmount = {
  /**
   * Builds the pair, or `null` when both members are absent. A half-supplied
   * pair throws: it is a defect at the call site, not a business outcome.
   */
  of(amount: Money | null | undefined, currency: Currency | null | undefined): OriginalAmount | null {
    const hasAmount = amount !== null && amount !== undefined;
    const hasCurrency = currency !== null && currency !== undefined;
    if (!hasAmount && !hasCurrency) return null;
    if (!hasAmount || !hasCurrency) {
      throw new InvalidTransactionError(
        'originalAmount is an all-or-nothing pair: supply both the amount and its currency, or neither — ' +
          'a lone amount is a number whose meaning depends on a currency nobody recorded',
      );
    }
    if (amount.currency.code !== currency.code) {
      throw new InvalidTransactionError(
        `originalAmount currency mismatch: the amount is denominated in ${amount.currency.code} but the pair declares ${currency.code}`,
      );
    }
    return Object.freeze({ amount, currency });
  },
};

/**
 * The canonical transaction.
 *
 * `version` is an optimistic-concurrency counter, incremented by exactly one
 * per accepted correction. It is what makes "update this transaction" safe
 * under two concurrent editors without either silently winning.
 */
export interface Transaction {
  readonly id: TransactionId;
  readonly tenantId: string;
  readonly userId: string;
  readonly accountRef: AccountRef;
  /** Exact, signed under the canonical convention. */
  readonly amount: Money;
  /** When the movement is booked to the account. Required — a transaction with no date is not a transaction. */
  readonly bookingDate: Date;
  /** When value is applied, where a source distinguishes it. Optional; never inferred from bookingDate. */
  readonly valueDate: Date | null;
  readonly merchant: HsfField | null;
  readonly description: HsfField;
  readonly note: HsfField | null;
  /** Source-stated amount in the source's own currency, all-or-nothing. */
  readonly originalAmount: OriginalAmount | null;
  readonly sourceKind: SourceKind;
  readonly status: TransactionStatus;
  readonly createdAt: Date;
  readonly version: number;
}

/**
 * Validates and freezes a transaction. Throws rather than returning a
 * `Result`: a malformed transaction reaching this constructor is a defect in
 * the calling use case, and every expected refusal (a bad date from a
 * request, an unknown currency code) is already a typed outcome upstream.
 */
export function createTransaction(fields: Transaction): Transaction {
  if (!Number.isInteger(fields.version) || fields.version < 1) {
    throw new InvalidTransactionError(
      `version must be a positive integer, got ${String(fields.version)}`,
    );
  }
  requireInstant('bookingDate', fields.bookingDate);
  requireInstant('createdAt', fields.createdAt);
  if (fields.valueDate !== null) requireInstant('valueDate', fields.valueDate);
  if (
    fields.originalAmount !== null &&
    fields.originalAmount.currency.code === fields.amount.currency.code
  ) {
    throw new InvalidTransactionError(
      `originalAmount is for a source currency that differs from the booked currency; ` +
        `both are ${fields.amount.currency.code}, which records nothing and invites a redundant second copy of the same figure`,
    );
  }
  return Object.freeze({ ...fields });
}

function requireInstant(field: string, value: Date): void {
  if (!(value instanceof Date) || Number.isNaN(value.getTime())) {
    throw new InvalidTransactionError(`${field} must be a valid Date`);
  }
}

/**
 * The fields a correction may change. Deliberately narrow: identity, owner,
 * account anchor, source kind, and creation instant are NOT correctable —
 * changing any of them would make the record a different record wearing the
 * same id, and the honest operation for that is a delete plus a new entry.
 */
export interface TransactionCorrection {
  readonly amount?: Money;
  readonly bookingDate?: Date;
  readonly valueDate?: Date | null;
  readonly merchant?: HsfField | null;
  readonly description?: HsfField;
  readonly note?: HsfField | null;
  readonly status?: TransactionStatus;
}

/**
 * Applies a correction, returning the next version. Pure — the caller pairs
 * the result with the `TransactionRevision` that records what changed, and
 * persists both in one transaction.
 */
export function applyCorrection(
  current: Transaction,
  correction: TransactionCorrection,
): Transaction {
  const next: Transaction = {
    ...current,
    ...(correction.amount !== undefined ? { amount: correction.amount } : {}),
    ...(correction.bookingDate !== undefined ? { bookingDate: correction.bookingDate } : {}),
    ...(correction.valueDate !== undefined ? { valueDate: correction.valueDate } : {}),
    ...(correction.merchant !== undefined ? { merchant: correction.merchant } : {}),
    ...(correction.description !== undefined ? { description: correction.description } : {}),
    ...(correction.note !== undefined ? { note: correction.note } : {}),
    ...(correction.status !== undefined ? { status: correction.status } : {}),
    version: current.version + 1,
  };
  if (correction.amount !== undefined && correction.amount.currency.code !== current.amount.currency.code) {
    throw new InvalidTransactionError(
      `a correction may not change the currency of a booked transaction (${current.amount.currency.code} -> ${correction.amount.currency.code}); ` +
        're-denominating a record in place would silently rewrite history, so the honest operation is a delete plus a new entry',
    );
  }
  return createTransaction(next);
}

File: apps/mobile/lib/features/transactions/domain/transaction.dart
// PURE DART ONLY. See lib/README.md — domain purity.
//
// ONE OF THE CALLER'S OWN TRANSACTIONS.
//
// `amount` is signed under the platform's canonical convention and
// `direction` restates it in words, so a screen renders an honest arrow
// without doing arithmetic on the sign. A person ENTERING an amount supplies a
// magnitude and a direction instead (`MoneyEntry`): a client that got the sign
// backwards would write a wrong financial record that looks exactly like a
// right one.
//
// `bookingDate` and `valueDate` are CALENDAR DAYS. They stay `CalendarDay` all
// the way to the widget — see `financial_accounts/domain/calendar_day.dart`
// for why turning one into a local-midnight `DateTime` moves a statement line
// across a month boundary.
//
// DELIBERATELY ABSENT, and to stay absent: the dedup fingerprint, its version
// and the occurrence ordinal; the import reference and the source row
// reference; any ciphertext or key version; and any category confidence —
// there is no score in this platform and none may be invented for display.
import 'package:meta/meta.dart';

import '../../financial_accounts/domain/calendar_day.dart';
import '../../financial_accounts/domain/money.dart';
import '../../financial_accounts/domain/source_rail.dart';

/// The transaction's own lifecycle.
enum TransactionStatus { posted, voided, unrecognised }

/// What the SOURCE itself said, before any mapping.
enum SourceDirection { debit, credit, notStated, unrecognised }

/// How the source's statement was turned into the canonical sign.
enum DirectionMapping {
  manualEntry,
  sourceDirectionWord,
  sourceSignedAmount,
  sourceSignedAmountInverted,
  unrecognised,
}

/// Who or what recorded a revision.
enum RevisionAttribution { sourceImport, manualEntry, userInput, unrecognised }

/// A field a correction may change.
enum RevisableField {
  amount,
  bookingDate,
  valueDate,
  merchant,
  description,
  note,
  status,
  unrecognised,
}

/// One transaction.
@immutable
final class Transaction {
  const Transaction({
    required this.transactionId,
    required this.accountId,
    required this.amount,
    required this.direction,
    required this.bookingDate,
    required this.valueDate,
    required this.eventOccurredAt,
    required this.sourceTimezone,
    required this.merchant,
    required this.description,
    required this.note,
    required this.originalAmount,
    required this.sourceKind,
    required this.availability,
    required this.status,
    required this.createdAt,
    required this.version,
  });

  final String transactionId;
  final String accountId;

  /// Signed, under the platform's convention. Never re-signed by this client.
  final Money amount;

  final MoneyDirection direction;

  /// The day the institution booked it. A calendar day, never an instant.
  final CalendarDay bookingDate;

  final CalendarDay? valueDate;

  /// A true instant, present only when the source stated one — and meaningful
  /// only with [sourceTimezone], which the platform requires alongside it.
  final DateTime? eventOccurredAt;

  final String? sourceTimezone;
  final String? merchant;
  final String description;
  final String? note;

  /// The amount as the source stated it, when its currency differed. It is
  /// shown BESIDE the booked amount and never instead of it: the two are in
  /// different currencies, and there is no rate on this surface to relate
  /// them.
  final Money? originalAmount;

  /// The rail this record arrived on.
  final SourceKind sourceKind;

  /// Whether this platform can run that rail. Carried beside the rail on every
  /// response, so a name is never mistaken for a working route.
  final RailAvailability availability;

  final TransactionStatus status;
  final DateTime createdAt;
  final int version;

  @override
  String toString() => 'Transaction($transactionId)';
}
